### PR TITLE
Refactor precommit to use actions/setup-python.

### PR DESCRIPTION
**Pull request by jagapiou**

### CODE REVIEW

File: .github/workflows/make_wheel.yml [hunk 1]
```diff
@@ -43,6 +43,7 @@ jobs:
         cfg:
           - { name: 'Linux LLVM+libstdc++', os: 'ubuntu-20.04', platform: 'manylinux_2_31_x86_64', cc: clang, cxx: clang++, config: }
           - { name: 'MacOS LLVM+libc++', os: 'macos-10.15', platform: 'macosx_10_15_x86_64', cc: clang, cxx: clang++, config: --config=libc++ }
+          - { name: 'MacOS LLVM+libc++', os: 'macos-11', platform: 'macosx_10_16_x86_64', cc: clang, cxx: clang++, config: --config=libc++ }
```
Comment (tkoeppe) [inline .github/workflows/make_wheel.yml]: Is this platform tag correct?

**Review by tkoeppe — CHANGES_REQUESTED** [review]

**Comment by tkoeppe** [comment]
If we build wheels with macos-11, will that make the wheels less compatible? Or are we expecting most users to be recent enough?

File: .github/workflows/make_wheel.yml [hunk 2]
```diff
@@ -43,6 +43,7 @@ jobs:
...
         py:
           - { version: '3.6', interp: 'cp36', abi: 'cp36m' }
```
Comment (tkoeppe) [inline .github/workflows/make_wheel.yml]: Does Py 3.6 still exist on macos-11?

**Review by tkoeppe — CHANGES_REQUESTED** [review]

File: .github/workflows/make_wheel.yml (hunk 1 above)
Comment (jagapiou) [reply .github/workflows/make_wheel.yml]: That's what gets built by the wheel target, and (in my own fork) I've confirmed this is used by `pip install`.

FYI: https://eclecticlight.co/2020/07/21/big-sur-is-both-10-16-and-11-0-its-official

File: .github/workflows/make_wheel.yml (hunk 2 above)
Comment (jagapiou) [reply .github/workflows/make_wheel.yml]: MacOS 11 defaults to Python 3.8. It's possible install 3.6 on OSX if desired.

**Comment by jagapiou** [comment]
I was thinking that we'd increase compatibility by building with both 11 and 10.15. But actually I've just checked and 11 can build from 10.15 wheels. I'll revert the last commit.

File: .github/workflows/make_wheel.yml (hunk 1 above)
Comment (tkoeppe) [reply .github/workflows/make_wheel.yml]: Ah right, that's the resulting ABI tag. Yeah, that makes sense.

File: .github/workflows/make_wheel.yml (hunk 2 above)
Comment (tkoeppe) [reply .github/workflows/make_wheel.yml]: Well, the question is whether this buil tinstruction as written will actually succeed on the GitHub action runner. The macos-11 documentation suggests that only versions 3.7 to 3.10 are available.

**Comment by tkoeppe** [comment]
> I was thinking that we'd increase compatibility by building with both 11 and 10.15. But actually I've just checked and 11 can build from 10.15 wheels. I'll revert the last commit.

The opposite would seem to be the case: the more recent your build environment, the more stringent your requirements, since you require users to be "at least as recent" as your build environment.

Yes, retaining the older env for building would seem to be better, not worse.

The only case you could make is for Py 3.10 support, I suppose. But maybe let's cross that bridge when there's active demand. (And we probably need to figure out a way to support M1 before we support Py 3.10.)

File: .github/workflows/precommit.yml [hunk 3]
```diff
@@ -29,7 +29,7 @@ jobs:
           - { name: 'Linux GCC', os: 'ubuntu-20.04', cc: gcc-10, cxx: g++-10, config: ''}
           - { name: 'Linux LLVM+libstdc++', os: 'ubuntu-20.04', cc: clang, cxx: clang++, config: ''}
           - { name: 'Linux LLVM+libc++', os: 'ubuntu-20.04', cc: clang, cxx: clang++, config: --config=libc++ }
-          - { name: 'MacOS LLVM+libc++', os: 'macos-10.15', cc: clang, cxx: clang++, config: --config=libc++ }
```
Comment (tkoeppe) [inline .github/workflows/precommit.yml]: Yes, we could, but do we have to or care? I'd prefer to leave this at the same version we use for wheel building.

Comment (jagapiou) [reply .github/workflows/precommit.yml]: Will revert.

File: .github/workflows/precommit.yml [hunk 4]
```diff
@@ -36,15 +36,19 @@ jobs:
...
-        run: |
-          sudo apt-get install python3-dev python3-numpy python3-packaging
-          sudo ln -s /usr/lib/llvm-10/include/c++/v1 /usr/include/c++/v1
-      - name: install infrastructure (MacOS)
-        if: matrix.os == 'macos-10.15'
-        run: sudo -H pip3 install numpy packaging
+
+      - uses: actions/setup-python@v3
+        with:
+          python-version: '3.10'
```
Comment (tkoeppe) [inline .github/workflows/precommit.yml]: I'm not sure we have a use for Py 3.10, do we? We just use the runner's default Python version. If we wanted to test multiple versions, we should set up a matrix like in make_wheel.yml, but I'm not entirely sure that's warranted. We don't have a lot of Python code, and so I'm not sure we benefit from another multiplicative increase of test configurations.

Comment (jagapiou) [reply .github/workflows/precommit.yml]: Changed to specify 3.x rather than 3.10.

**Comment by tkoeppe** [comment]
I've updated the precommit to use macos-12 and macos-13, and make wheels for pythons 3.8-3.11, and I'm not sure there's much else we need to do at the moment.

What I did notice for macOS was that there's no macOS equivalent of Linux's "manylinux" configuration, so we don't have a single, universal macOS wheel. I've heard people specify compiler flags to build binaries compatible with 10.15 and newer, which is perhaps something we could entertain. I'm not sure what the downsides are compared to building for each platform separately.

But in any case I'm going to close this. Please feel free to reopen if there is new information.

**Pull request closed without merging**